### PR TITLE
Backport 2.x: fix SHA384 guards in TLS

### LINKED ISSUE #4472: Change SHA1 define guard to SHA512 in ssl_ciphersuites.c
_Originally posted by @gilles-peskine-arm in https://github.com/ARMmbed/mbedtls/pull/4304#discussion_r625742279_

in library/ssl_ciphersuites.c (https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_ciphersuites.c#L854), there is a mistake in the `#if defined` code guard for SHA1. #4304 for the 3.0 release will change it to SHA384. In 2.16 and 2.2x this define guard should be for SHA512.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Even after the addition of `MBEDTLS_SHA512_NO_SHA384`, the TLS code gated the availability of SHA384 ciphersuites on `MBEDTLS_SHA512_C` alone, and in one case due to a typo on `MBEDTLS_SHA1_C`. Fix #4472, #4499.

For 3.0 this is fixed in https://github.com/ARMmbed/mbedtls/pull/4304 (the issues were discovered while reviewing that PR).

I have not fixed a similar issue in `ssl_cookie.c` because I'm not sure what the fix is and it remains partly open in 3.0. It is filed as https://github.com/ARMmbed/mbedtls/issues/4501.

There are no non-regression tests because I don't think it's easy to add them. I've filed the lack of tests as #4500. To test manually, build `development_2.x` with `MBEDTLS_SHA512_NO_SHA384` defined and observe that `programs/ssl/ssl_client2 help` lists SHA384 cipher suites but they do not work. With this patch, the ciphersuites are no longer offered.

### CODE REVIEW

File: library/ssl_ciphersuites.c [hunk 1]
```diff
@@ -887,7 +887,7 @@ static const mbedtls_ssl_ciphersuite_t ciphersuite_definitions[] =
       0 },
 #endif /* MBEDTLS_SHA256_C */
 
-#if defined(MBEDTLS_SHA1_C)
+#if defined(MBEDTLS_SHA512_C)
```
Comment (ronald-cron-arm) [inline library/ssl_ciphersuites.c]: Change also to MBEDTLS_SHA512_C in the #endif comment below.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM. Just spotted a very minor miss.

File: library/ssl_ciphersuites.c [hunk 2]
```diff
@@ -887,13 +892,13 @@ static const mbedtls_ssl_ciphersuite_t ciphersuite_definitions[] =
...
 
-#if defined(MBEDTLS_SHA1_C)
+#if defined(HAVE_SHA384)
     { MBEDTLS_TLS_RSA_WITH_CAMELLIA_256_GCM_SHA384, "TLS-RSA-WITH-CAMELLIA-256-GCM-SHA384",
       MBEDTLS_CIPHER_CAMELLIA_256_GCM, MBEDTLS_MD_SHA384, MBEDTLS_KEY_EXCHANGE_RSA,
       MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3,
       MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3,
       0 },
-#endif /* MBEDTLS_SHA1_C */
+#endif /* MBEDTLS_SHA384_C */
```
Comment (ronald-cron-arm) [inline library/ssl_ciphersuites.c:901]:
```suggestion
#endif /* HAVE_SHA384 */
```

Comment (gilles-peskine-arm) [reply library/ssl_ciphersuites.c]: Ok, I think I got it right now (again, I amended the commit).

I wish we had an automatic check on #endif guards. Unfortunately I haven't found any easy-to-integrate tool for that.

**Review by ronald-cron-arm — APPROVED** [review]
All good to me now. Thanks.

**Review by daverodgman — APPROVED** [review]
This looks good

**Pull request merged**